src/main/java/uk/gov/justice/hmpps/prison/web/handler/ControllerAdvice.kt

```diff
@@ -0,0 +1,389 @@
+          .build(),
+      )
+  }
+
+  @ExceptionHandler(NoContentException::class)
+  fun handleNoContentException(e: NoContentException): ResponseEntity<ErrorResponse> {
+    log.debug("No content (204) returned with message {}", e.message)
+    return ResponseEntity
+      .status(HttpStatus.NO_CONTENT)
+      .body(
+        ErrorResponse
+          .builder()
+          .userMessage("No content returned")
+          .status(HttpStatus.NO_CONTENT.value())
+          .developerMessage(e.message)
+          .build(),
+      )
+  }
+
+  @ExceptionHandler(MethodArgumentNotValidException::class)
+  fun handleMethodArgumentNotValidException(e: MethodArgumentNotValidException): ResponseEntity<ErrorResponse> {
+    val errors = e.bindingResult.fieldErrors
+      .stream()
+      .map { error: FieldError -> "Field: " + error.field + " - " + error.defaultMessage }
+      .collect(Collectors.joining(", "))
+    return ResponseEntity
+      .status(HttpStatus.BAD_REQUEST)
+      .body(
+        ErrorResponse
+          .builder()
+          .userMessage(errors)
+          .status(HttpStatus.BAD_REQUEST.value())
+          .developerMessage(errors)
+          .build(),
+      )
+  }
+
+  @ExceptionHandler(Exception::class)
+  fun handleException(e: Exception): ResponseEntity<ErrorResponse> {
+    log.error("Unexpected exception", e)
+    return ResponseEntity
+      .status(HttpStatus.INTERNAL_SERVER_ERROR)
+      .body(
+        ErrorResponse
+          .builder()
+          .userMessage(e.message)
+          .status(HttpStatus.INTERNAL_SERVER_ERROR.value())
+          .developerMessage(e.message)
+          .build(),
+      )
+  }
+
+  @ExceptionHandler(BadRequestException::class)
+  fun handleServiceException(e: BadRequestException): ResponseEntity<ErrorResponse> {
+    log.debug("Bad Request (400) returned with message {}", e.message)
+    return ResponseEntity
+      .status(HttpStatus.BAD_REQUEST)
+      .body(
+        ErrorResponse
+          .builder()
+          .userMessage(e.message)
+          .status(HttpStatus.BAD_REQUEST.value())
+          .errorCode(e.errorCode)
+          .developerMessage(e.message)
+          .build(),
+      )
+  }
+
+  @ExceptionHandler(DuplicateKeyException::class)
+  fun handleServiceException(e: DuplicateKeyException): ResponseEntity<ErrorResponse> {
+    log.debug("Conflict (409) returned with message {}", e.message)
+    return ResponseEntity
+      .status(HttpStatus.CONFLICT)
+      .body(
+        ErrorResponse
+          .builder()
+          .userMessage(e.message)
+          .status(HttpStatus.CONFLICT.value())
+          .developerMessage(e.message)
+          .build(),
+      )
+  }
+
+  @ExceptionHandler(JpaSystemException::class)
+  fun handleJpaException(e: JpaSystemException): ResponseEntity<ErrorResponse> {
+    try {
+      if (e.cause is GenericJDBCException && (e.cause as GenericJDBCException).cause is SQLException) {
+        return handleJpaSqlException(e, e.cause!!.cause!!)
```

Comment:
pass through `e.cause.cause` as we've already checked they're not null (as they have a type)